modules/improvedigitalBidAdapter.js

```diff
@@ -29,43 +15,169 @@ const IFRAME_SYNC_URL = 'https://hb.360yield.com/prebid-universal-creative/load-
 
 const VIDEO_PARAMS = {
   DEFAULT_MIMES: ['video/mp4'],
-  SUPPORTED_PROPERTIES: ['mimes', 'minduration', 'maxduration', 'protocols', 'w', 'h', 'startdelay', 'placement', 'linearity', 'skip', 'skipmin',
-    'skipafter', 'sequence', 'battr', 'maxextended', 'minbitrate', 'maxbitrate', 'boxingallowed', 'playbackmethod', 'playbackend', 'delivery', 'pos', 'companionad',
-    'api', 'companiontype', 'ext'],
   PLACEMENT_TYPE: {
     INSTREAM: 1,
     OUTSTREAM: 3,
   }
 };
 
-const NATIVE_DATA = {
-  VERSION: '1.2',
-  ASSET_TYPES: {
-    TITLE: 'title',
-    IMG: 'img',
-    DATA: 'data',
+export const CONVERTER = ortbConverter({
+  context: {
+    nativeRequest: {
+      eventtrackers: [
+        {event: 1, methods: [1, 2]},
+      ]
+    }
+  },
+  imp(buildImp, bidRequest, context) {
+    const imp = buildImp(bidRequest, context);
+    imp.secure = Number(window.location.protocol === 'https:');
+    if (imp.hasOwnProperty('instl')) {
+      // accept booleans for 'instl'
+      // TODO: does this make sense? we expect ortb2Imp to follow the ORTB spec
+      imp.instl = imp.instl ? 1 : 0
+    }
+    if (!imp.bidfloor && bidRequest.params.bidFloor) {
+      imp.bidfloor = bidRequest.params.bidFloor;
+      imp.bidfloorcur = getBidIdParameter('bidFloorCur', bidRequest.params).toUpperCase() || 'USD'
+    }
+    const bidderParamsPath = context.extendMode ? 'ext.prebid.bidder.improvedigital' : 'ext.bidder';
+    const placementId = bidRequest.params.placementId;
+    if (placementId) {
+      deepSetValue(imp, `${bidderParamsPath}.placementId`, placementId);
+      if (context.extendMode) {
+        deepSetValue(imp, 'ext.prebid.storedrequest.id', '' + placementId);
+      }
+    } else {
+      deepSetValue(imp, `${bidderParamsPath}.publisherId`, getBidIdParameter('publisherId', bidRequest.params));
+      deepSetValue(imp, `${bidderParamsPath}.placementKey`, getBidIdParameter('placementKey', bidRequest.params));
+    }
+    deepSetValue(imp, `${bidderParamsPath}.keyValues`, getBidIdParameter('keyValues', bidRequest.params) || undefined);
+
+    // Adding GPID
+    // TODO: this logic is in modules/gptPreAuction; should it be in core?
+    deepSetValue(imp, 'ext.gpid',
+      imp.ext?.gpid ||
+      deepAccess(bidRequest, 'ortb2Imp.ext.data.pbadslot') ||
+      deepAccess(bidRequest, 'ortb2Imp.ext.data.adserver.adslot')
+    );
+
+    return imp;
+  },
+  request(buildRequest, imps, bidderRequest, context) {
+    const request = buildRequest(imps, bidderRequest, context);
+    mergeDeep(request, {
+      id: getUniqueIdentifierStr(),
+      source: {
+        // TODO: it may be better to not set this if there's more than one bidRequest - see https://github.com/prebid/Prebid.js/issues/8543
```

Comment:
I don't see any harm. According to imp.ext.tid spec if both source.tid and imp.ext.tid are provided then imp.ext.tid should override, with source.tid used as a fallback for any items in the impression array that don't specify a imp.ext.tid.

Reply:
if only source.tid is specified, it should be taken as the default value for imp.ext.tid (according to that spec). That means that it's never correct to set it if there is more than one imp in the request.

Reply:
Agreed - the matter is not as settled as I thought.